File: 2026/leetcode/17.letter-combinations-of-a-phone-number.java
/*
 * @lc app=leetcode.cn id=17 lang=java
 * @lcpr version=30204
 *
 * [17] 电话号码的字母组合
 */


// @lcpr-template-start

// @lcpr-template-end
// @lc code=start

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.List;;

class Solution {
    Map<Character,List<Character>>  map = new HashMap<>();

    public List<String> letterCombinations(String digits) {
        map.put('2', Arrays.asList('a','b','c'));
        map.put('3', Arrays.asList('d','e','f'));
        map.put('4', Arrays.asList('g','h','i'));
        map.put('5', Arrays.asList('j','k','l'));
        map.put('6', Arrays.asList('m','n','o'));
        map.put('7', Arrays.asList('p','q','r','s'));
        map.put('8', Arrays.asList('t','u','v'));
        map.put('9', Arrays.asList('w','x','y','z'));
       return letterCombinationsWithIndex(digits,0);
    }

    public List<String> letterCombinationsWithIndex(String digits, int index){
        if(index == digits.length()){
            return Arrays.asList("");
        }
        List<String> res = new ArrayList<>(); 
        List<Character > l = map.get(digits.charAt(index));
         List<String> xx = letterCombinationsWithIndex(digits,index+1);
        for (int i =  0 ; i < l.size(); i++){
            char c = l.get(i);
            for (int j = 0 ; j< xx.size() ;j++){
                res.add(c + xx.get(j));
            }
        }
        return res ;
    } 
}
// @lc code=end



/*
// @lcpr case=start
// "23"\n
// @lcpr case=end

// @lcpr case=start
// "2"\n
// @lcpr case=end

 */

File: 2026/leetcode/17.letter-combinations-of-a-phone-number.go
/*
 * @lc app=leetcode.cn id=17 lang=golang
 * @lcpr version=30204
 *
 * [17] 电话号码的字母组合
 */

// @lcpr-template-start

// @lcpr-template-end
// @lc code=start
// 核心思路：把每个数字映射到候选字母，递归/回溯拼接每一位的所有组合。
// 关键点：递归到末尾时返回空串作为拼接基底，再逐层组合当前字符和后续结果。
// 复杂度：时间 O(4^n * n)，空间 O(4^n * n) 用于结果集。
func letterCombinations(digits string) []string {
	mm := map[rune][]string{
		'2': {"a", "b", "c"},
		'3': {"d", "e", "f"},
		'4': {"g", "h", "i"},
		'5': {"j", "k", "l"},
		'6': {"m", "n", "o"},
		'7': {"p", "q", "r", "s"},
		'8': {"t", "u", "v"},
		'9': {"w", "x", "y", "z"},
	}
	return letterCombinationsWithIndex(mm, []rune(digits), 0)
}
func letterCombinationsWithIndex(mm map[rune][]string, digits []rune, index int) []string {
	if index == len(digits) {
		return []string{""}
	}
	res := make([]string, 0)
	ss := mm[digits[index]]
	next := letterCombinationsWithIndex(mm, digits, index+1)
	for i := 0; i < len(ss); i++ {
		o := ss[i]
		for j := 0; j < len(next); j++ {
			res = append(res, o+next[j])
		}
	}
	return res
}

// @lc code=end

/*
// @lcpr case=start
// "23"\n
// @lcpr case=end

// @lcpr case=start
// "2"\n
// @lcpr case=end

*/

